### PR TITLE
Fix issues #24,44 -> 'indexing array with string' problem (also check Promise.serial() input)

### PR DESCRIPTION
**Pull request by smittytone**
Resolve issues #44, #24; update copyright date and owner on lib, licence and readme

### CODE REVIEW

File: LICENSE [hunk 1]
```diff
@@ -1,5 +1,6 @@
 MIT License
 
+Copyright 2021 Twilio
```
Comment (rlipscombe) [inline LICENSE]: Ideally copyright and version changes would be in a separate commit from the fix.

File: Promise.lib.nut [hunk 2]
```diff
@@ -128,17 +129,15 @@ class Promise {
     // @return {boolean}
     //
     function _isPromise(value) {
-        if (
-            // detect that the value is some form of Promise
+        if (// detect that the value is some form of Promise
```
Comment (rlipscombe) [inline Promise.lib.nut:133]: Ditto with the reformatting changes.

File: Promise.lib.nut [hunk 3]
```diff
@@ -240,21 +239,30 @@ class Promise {
...
         local i = 0;
         return loop(
-            @() i < promises.len(),
+            @() i < promises.len(),     // This is a Squirrel lamda --
+                                        // see https://developer.electricimp.com/squirrel/squirrel-guide/functions#lambda-functions
             function () {
-                return "function" == type(promises[i])
-                     ? promises[i++]()
-                     : promises[i++];
+                local res = promises[i++];
```
Comment (rlipscombe) [inline Promise.lib.nut]: Suggest a comment right here explaining the problem, so that no-one's tempted to refactor it back to the broken behaviour.

File: Promise.lib.nut [hunk 4]
```diff
@@ -240,21 +239,30 @@ class Promise {
...
+                if ("function" == typeof res) res = res();
+                return res;
```
Comment (rlipscombe) [inline Promise.lib.nut]: I don't like this. Prefer the following:
```
if ("function" == typeof res) {
    return res();
} else {
    return res;
}
```

Reasoning: (1) `if` followed by the action on the same line is harder to read; (2) reusing `res` -- don't like; (3) braces are always good; (4) early return is fine, imo.

Comment (rlipscombe) [reply Promise.lib.nut]: On that note, `res` isn't a great name for this.

Comment (smittytone) [reply Promise.lib.nut]: Done

File: tests/Serial.agent.test.nut [hunk 5]
```diff
@@ -143,11 +143,17 @@ class SerialTestCase extends ImpTestCase {
 
     function testSerialWithArrayCallFix() {
 
-        local executeCommand = Promise(function(resolve, reject) {
-            imp.wakeup(0, function() {
-                resolve("Array function called");
+        // TEST FIX TO PROMISE.SERIAL
+        // See https://forums.electricimp.com/t/my-integer-suddenly-becomes-a-string/6454/7
+
+        local executeCommand = function() {
```
Comment (pdh11) [inline tests/Serial.agent.test.nut]: Does this actually test for the bug, i.e. does it fail if run against the unfixed version of the code? I'd expect to see either a write to a non-local variable, or an explicit assertion on type(this)...

Comment (smittytone) [reply tests/Serial.agent.test.nut]: That seems very reasonable. Code updated. Tests fail with old lib, pass with new. BUT... this involves such an imbroglio of functions that return promises and promises themselves that I can't honestly say I'm 100% sure the test pass is due to a sensible check for success. At least it is not failing with 'indexing array with string'. Thoughts?

Comment (pdh11) [reply tests/Serial.agent.test.nut]: Failing with the old lib, passing with the new is the acid test here I think. Yes it's a little complex, but any passer-by who is confused by it can follow the forums link, and/or review the related commits, and read all the gory details. There's already a decent comment on Promise.serial which mentions the actual issue.

Comment (smittytone) [reply tests/Serial.agent.test.nut]: Sounds good. Thank you!
